crypto/hash/hash_test.go

```diff
@@ -188,6 +199,29 @@ func TestSha3(t *testing.T) {
 	})
 }
 
+// TestKeccak is a specific test of Keccak-256.
+// It compares the hashes of random data of different lengths to
+// the output of Go LegacyKeccak.
+func TestKeccak(t *testing.T) {
+	r := time.Now().UnixNano()
+	rand.Seed(r)
+	t.Logf("math rand seed is %d", r)
+	var expected [HashLenKeccak_256]byte
+
+	for i := 0; i < 5000; i++ {
+		value := make([]byte, i)
+		rand.Read(value)
+		k := sha3.NewLegacyKeccak256()
+		k.Write(value)
+		k.Sum(expected[:0])
```

Comment:
The documentation of `Sum`:
> Sum appends the current hash to b and returns the resulting slice.
> It does not change the underlying hash state.

To me, it looks like you are pre-allocating the slice based on your knowledge of how long you are expecting the result to be. To me, this is a bit confusing. This is a test and therefore probably not performance critical. I would suggest favouring readability over runtime speed:

```suggestion
		expectedHash := k.Sum(nil)
```